### PR TITLE
bpo-47053: Refactor BINARY_OP_INPLACE_ADD_UNICODE

### PR DESCRIPTION
**Pull request by sweeneyde**
Letting `PyUnicode_Append` directly modify the target local avoids the next `STORE_FAST`, and it avoids the `GETLOCAL(next_oparg) = NULL;`.

https://bugs.python.org/issue47053

### CODE REVIEW

**Comment by sweeneyde** [comment]
Microbenchmark:
```
./python -m pyperf timeit -s "from itertools import repeat; r = 'a'*10_000; s=''" "for c in r: s += c"
```

Mean +- std dev: [before] 454 us +- 3 us -> [after] 437 us +- 4 us: 1.04x faster

File: Python/ceval.c [hunk 1]
```diff
@@ -2003,28 +2003,32 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
...
-            /* In the common case, there are 2 references to the value
-            * stored in 'variable' when the v = v + ... is performed: one
-            * on the value stack (in 'v') and one still stored in the
-            * 'variable'.  We try to delete the variable now to reduce
-            * the refcnt to 1.
-            */
-            PyObject *var = GETLOCAL(next_oparg);
-            DEOPT_IF(var != left, BINARY_OP);
+            PyObject **target_local = &GETLOCAL(_Py_OPARG(true_next));
+            DEOPT_IF(*target_local != left, BINARY_OP);
```
Comment (Fidget-Spinner) [inline Python/ceval.c:2009]: When does this happen?

**Review by Fidget-Spinner — APPROVED** [review]
LGTM.

File: Python/ceval.c (hunk 1 above)
Comment (sweeneyde) [reply Python/ceval.c:2009]: On the last iteration here for example:
```
>>> pairs = [('foo',) * 2 for _ in range(10_000)]
>>> pairs.append(('foo', 'bar'))
>>> def f():
...     for x, y in pairs:
...         x = y + 'a'
...
>>> f()
```

**Pull request merged**